### PR TITLE
expression: use parameter ctx to do evaluate instead of the inner one for all builtinFuncs

### LINKED ISSUE #48408: Use the parameter `ctx` instead of inner `b.ctx` in `evalXXX` methods in `builtinFunc`

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #48408

### What is changed and how it works?

This is the second step to remove inner ctx in `ScalarFunction`, yo can see detail in: #47958

In this PR, we:

1. remove the inner `ctx` in `baseBuiltinFunc` and move it outside to `ScalarFunction` instead.
2. `builtinFuncNew` is no longer needed any more, just remove it

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/executor/aggfuncs/builder.go [hunk 1]
```diff
@@ -457,7 +457,7 @@ func buildGroupConcat(ctx sessionctx.Context, aggFuncDesc *aggregation.AggFuncDe
 	default:
 		// The last arg is promised to be a not-null string constant, so the error can be ignored.
 		c, _ := aggFuncDesc.Args[len(aggFuncDesc.Args)-1].(*expression.Constant)
-		sep, _, err := c.EvalString(nil, chunk.Row{})
+		sep, _, err := c.EvalString(ctx, chunk.Row{})
```
Comment (lcwangchao) [inline pkg/executor/aggfuncs/builder.go:460]: Make sure `EvalString` not panic. But I don't know why the previous code used `nil` as the input parameter. @XuHuaiyu PTAL

File: pkg/planner/core/expression_rewriter.go [hunk 2]
```diff
@@ -1846,7 +1846,7 @@ func (er *expressionRewriter) patternLikeOrIlikeToExpression(v *ast.PatternLikeO
 	isPatternExactMatch := false
 	// Treat predicate 'like' or 'ilike' the same way as predicate '=' when it is an exact match and new collation is not enabled.
 	if patExpression, ok := er.ctxStack[l-1].(*expression.Constant); ok && !collate.NewCollationEnabled() {
-		patString, isNull, err := patExpression.EvalString(nil, chunk.Row{})
+		patString, isNull, err := patExpression.EvalString(er.sctx, chunk.Row{})
```
Comment (lcwangchao) [inline pkg/planner/core/expression_rewriter.go:1849]: Make sure `EvalString` not panic. PTAL @winoros

**Review by YangKeao — APPROVED** [review]
LGTM

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**